File: python/techai_webutils/src/techai_webutils/core/interfaces/store.py
"""Store interface for infrastructure-level data access.

Mirrors Go's ``interfaces.Store[T, P, ID]``. Same CRUD surface as
Repository but represents the infrastructure boundary (e.g., asyncpg,
SQLAlchemy) rather than the business-logic boundary.
"""

from __future__ import annotations

from abc import ABC, abstractmethod
from typing import TypeVar, TYPE_CHECKING


if TYPE_CHECKING:
    from techai_webutils.core.domain_types.types import Page, PageRequest

T = TypeVar("T")
P = TypeVar("P")
ID = TypeVar("ID")


class Store[T, P, ID](ABC):
    """Low-level data access interface.

    T = entity type, P = query/filter params, ID = identifier type.
    """

    @abstractmethod
    async def get(self, entity_id: ID) -> T:
        """Retrieve an entity by ID."""
        ...

    @abstractmethod
    async def list(self, params: P, page: PageRequest) -> Page[T]:
        """Retrieve entities matching parameters with pagination."""
        ...

    @abstractmethod
    async def create(self, entity: T) -> T:
        """Persist a new entity and return it with generated fields populated."""
        ...

    @abstractmethod
    async def update(self, entity_id: ID, entity: T) -> T:
        """Modify an existing entity identified by id."""
        ...

    @abstractmethod
    async def delete(self, entity_id: ID) -> None:
        """Remove an entity by ID."""
        ...

    @abstractmethod
    async def exists(self, entity_id: ID) -> bool:
        """Check if an entity with the given ID exists."""
        ...

File: go/core/interfaces/store.go
package interfaces

// Store defines the low-level data access interface that concrete database
// adapters implement. It has the same CRUD surface as Repository but represents
// the infrastructure boundary (e.g., Ent, sqlc, pgx) rather than the
// business-logic boundary. It is composed from the role interfaces,
// so a full-CRUD adapter embeds the whole set while a read-only or append-only
// adapter can embed only the roles it honors.
//
// T = entity type, P = query/filter params, ID = identifier type.
//
// Phase 1: Interface definition.
// Phase 3: Ent and sqlc implementations.
type Store[T any, P any, ID comparable] interface {
	// Reader provides Get (retrieve by id).
	Reader[T, ID]
	// Lister provides List (paginated query by params).
	Lister[T, P]
	// Writer provides Create and Update.
	Writer[T, ID]
	// Deleter provides Delete (remove by id).
	Deleter[ID]
	// Exister provides Exists (existence check by id).
	Exister[ID]
}
